### PR TITLE
Implement `BundleRepo` via RocksDB

### PR DESCRIPTION
**Pull request by kettlebell**

### CODE REVIEW

File: spectrum-offchain-lm/src/data/bundle.rs [hunk 1]
```diff
@@ -140,24 +183,18 @@ impl IntoBoxCandidate for StakingBundle {
 impl TryFromBox for StakingBundle {
     fn try_from_box(bx: ErgoBox) -> Option<StakingBundle> {
         if let Some(ref tokens) = bx.tokens {
-            if tokens.len() == 2 && bx.ergo_tree == bundle_validator() {
+            if tokens.len() == 3 && bx.ergo_tree == bundle_validator() {
```
Comment (kettlebell) [inline spectrum-offchain-lm/src/data/bundle.rs:186]: Reflecting new changes in https://github.com/spectrum-finance/ergo-dex/blob/67a4d8d4dfb1522131a233befb37da805cba8299/contracts/lqmining/simple/StakingBundle.sc

File: spectrum-offchain-lm/src/validators.rs [hunk 2]
```diff
@@ -38,18 +38,7 @@ pub fn redeem_validator_temp() -> Vec<u8> {
...
-        03938602720a730cb27211730d00938c7212018c720b019399720e8c72120299720e9c72157
-        20d93b27211730e00720ced938c7216018cb27205730f0001927e8c721602069d9c9c7e9de4
-        c6720405057e721305067e720d067e999d720e720d7215067e720606958f7207731093b2db6
-        308b2a47311007312008602720a73137314";
+    let raw = "19ab0315040004000404040404040402040004000500040204020400050204040400040205000404040005fcffffffffffffffff010100d80ed601b2a5730000d602db63087201d603e4c6a7050ed604b2a4730100d605db63087204d6068cb2720573020002d607998cb27202730300027206d608e4c6a7040ed609db6308a7d60a8cb2720973040001d60bb27209730500d60cb27209730600d60d8c720c02d60e8c720b02d1ed938cb27202730700017203959372077308d808d60fb2a5e4e3000400d610b2a5e4e3010400d611db63087210d612b27211730900d613b2e4c672040410730a00d614c672010804d6157e99721395e67214e47214e4c67201070405d616b2db6308720f730b00eded93c2720f7208ededededed93e4c67210040e720893e4c67210050e7203938602720a730cb27211730d00938c7212018c720b019399720e8c72120299720e9c7215720d93b27211730e00720ced938c7216018cb27205730f0001927e8c721602069d9c9c7e9de4c6720405057e721305067e720d067e999d720e720d7215067e720606958f7207731093b2db6308b2a47311007312008602720a73137314";
```
Comment (kettlebell) [inline spectrum-offchain-lm/src/validators.rs:41]: Need this string on a single-line in Rust.

File: spectrum-offchain-lm/src/data/bundle.rs [hunk 3]
```diff
@@ -140,24 +183,18 @@ impl IntoBoxCandidate for StakingBundle {
...
                 let redeemer_prop = ErgoTree::sigma_parse_bytes(
-                    &*bx.additional_registers
-                        .get(NonMandatoryRegisterId::R4)?
-                        .as_option_constant()?
+                    &bx.get_register(NonMandatoryRegisterId::R4.into())?
```
Comment (kettlebell) [inline spectrum-offchain-lm/src/data/bundle.rs:188]: Trying to work on `additional_registers` field directly crashes. Better to work with `get_register` method directly (`sigma-rust` should actually be updated here, make `additional_registers` non-pub).

File: spectrum-offchain-lm/src/bundle.rs [hunk 4]
```diff
@@ -4,34 +4,345 @@ use async_trait::async_trait;
...
 use crate::data::{AsBox, BundleId, BundleStateId, PoolId};
 
 pub mod process;
+pub mod rocksdb;
 
 #[async_trait]
 pub trait BundleRepo {
-    /// Select bundles corresponding to
+    fn get_epoch_length(&self) -> u32;
+    fn get_height_range(&self, epoch_ix: u32) -> EpochHeightRange;
```
Comment (oskin1) [inline spectrum-offchain-lm/src/bundle.rs:18]: Those may be different for each `pool_id`, we store bundles for all pools in the network in this Repo.

File: spectrum-offchain-lm/src/bundle/rocksdb.rs [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            let mut bundle_ids = vec![];
+            for (_, value) in db
+                .iterator(rocksdb::IteratorMode::From(
+                    &bincode::serialize(STATE_PREFIX.as_bytes()).unwrap(),
+                    rocksdb::Direction::Forward,
+                ))
+                .flatten()
+            {
+                if let Ok(bundle) = bincode::deserialize::<'_, AsBox<StakingBundle>>(&value) {
+                    let bundle_height = bundle.0.creation_height;
```
Comment (oskin1) [inline spectrum-offchain-lm/src/bundle/rocksdb.rs:51]: creation_height of the bundle box has nothing to do with epochs bundle is eligible for

**Review by oskin1 — APPROVED** [review]

**Pull request merged**